File: WhatsApp-agents/baileys/src/contracts/outbound.ts
//Mariano Montini ('bosque', 'bosquestudio')
import { z } from 'zod'

// Text outbound input - destination WhatsApp number and plain body.
export const textOutboundInputSchema = z.object({
  to: z.string().min(1),
  body: z.string().min(1),
})
export type TextOutboundInput = z.infer<typeof textOutboundInputSchema>

// Interactive button definition - id and visible title (unstable on Baileys).
export const interactiveButtonSchema = z.object({
  id: z.string().min(1),
  title: z.string().min(1).max(20),
})

// Interactive buttons input - header/body/footer plus up to three buttons.
export const interactiveButtonsInputSchema = z.object({
  to: z.string().min(1),
  bodyText: z.string().min(1),
  buttons: z.array(interactiveButtonSchema).min(1).max(3),
  headerText: z.string().optional(),
  footerText: z.string().optional(),
})
export type InteractiveButtonsInput = z.infer<typeof interactiveButtonsInputSchema>

// List row definition - id, title, optional description for list messages.
export const listRowSchema = z.object({
  id: z.string().min(1),
  title: z.string().min(1).max(24),
  description: z.string().max(72).optional(),
})

// Interactive list input - sectioned list (documented optional / unstable).
export const interactiveListInputSchema = z.object({
  to: z.string().min(1),
  bodyText: z.string().min(1),
  buttonText: z.string().min(1).max(20),
  sections: z
    .array(
      z.object({
        title: z.string().min(1).max(24),
        rows: z.array(listRowSchema).min(1),
      }),
    )
    .min(1),
  headerText: z.string().optional(),
  footerText: z.string().optional(),
})
export type InteractiveListInput = z.infer<typeof interactiveListInputSchema>

// Media outbound input - image or document with link and optional caption/filename.
export const mediaOutboundInputSchema = z.object({
  to: z.string().min(1),
  type: z.enum(['image', 'document']),
  link: z.string().url(),
  caption: z.string().optional(),
  filename: z.string().optional(),
})
export type MediaOutboundInput = z.infer<typeof mediaOutboundInputSchema>

// Outbound message kinds - discriminated union for adapter + humanizer.
export type OutboundMessage =
  | { kind: 'text'; to: string; text: string }
  | {
      kind: 'buttons'
      to: string
      title?: string
      description: string
      footer?: string
      buttons: Array<{ id: string; displayText: string }>
    }
  | {
      kind: 'list'
      to: string
      title?: string
      description: string
      buttonText: string
      footer?: string
      sections: Array<{
        title: string
        rows: Array<{ id: string; title: string; description?: string }>
      }>
    }
  | {
      kind: 'media'
      to: string
      mediatype: 'image' | 'document'
      media: string
      caption?: string
      fileName?: string
    }

File: WhatsApp-agents/whatsmeow/internal/contracts/outbound.go
package contracts

import (
	"errors"
	"net/url"
	"strings"
)

// OutboundMessage - discriminated kinds for adapter + humanizer.
type OutboundMessage struct {
	Kind        string
	To          string
	Text        string
	Title       string
	Description string
	Footer      string
	ButtonText  string
	Buttons     []OutboundButton
	Sections    []OutboundSection
	MediaType   string
	Media       string
	Caption     string
	FileName    string
}

type OutboundButton struct {
	ID          string
	DisplayText string
}

type OutboundSection struct {
	Title string
	Rows  []OutboundRow
}

type OutboundRow struct {
	ID          string
	Title       string
	Description string
}

type TextOutboundInput struct {
	To   string
	Body string
}

type MediaOutboundInput struct {
	To       string
	Type     string
	Link     string
	Caption  string
	Filename string
}

type InteractiveButtonsInput struct {
	To         string
	BodyText   string
	Buttons    []struct{ ID, Title string }
	HeaderText string
	FooterText string
}

type InteractiveListInput struct {
	To         string
	BodyText   string
	ButtonText string
	HeaderText string
	FooterText string
	Sections   []struct {
		Title string
		Rows  []struct {
			ID          string
			Title       string
			Description string
		}
	}
}

var (
	errEmptyTo   = errors.New("to must be non-empty")
	errEmptyBody = errors.New("body must be non-empty")
	errMediaURL  = errors.New("media link must be a valid URL")
)

// ValidateTextInput - destination and non-empty body.
func ValidateTextInput(input TextOutboundInput) (TextOutboundInput, error) {
	input.To = strings.TrimSpace(input.To)
	input.Body = strings.TrimSpace(input.Body)
	if input.To == "" {
		return input, errEmptyTo
	}
	if input.Body == "" {
		return input, errEmptyBody
	}
	return input, nil
}

// ValidateMediaInput - image or document with HTTPS/HTTP URL.
func ValidateMediaInput(input MediaOutboundInput) (MediaOutboundInput, error) {
	input.To = strings.TrimSpace(input.To)
	input.Type = strings.TrimSpace(input.Type)
	input.Link = strings.TrimSpace(input.Link)
	if input.To == "" {
		return input, errEmptyTo
	}
	if input.Type != "image" && input.Type != "document" {
		return input, errors.New("media type must be image or document")
	}
	if _, err := url.ParseRequestURI(input.Link); err != nil {
		return input, errMediaURL
	}
	return input, nil
}
